### PR TITLE
Simplify api

### PR DESCRIPTION
**Pull request by reciprocated**
WIP on #7 #15

### CODE REVIEW

File: trlx/model/accelerate_ilql_model.py [hunk 1]
```diff
@@ -1,96 +1,56 @@
...
-        timer = Clock()
-
-        eos_token_id = self.tokenizer.eos_token_id if self.tokenizer else 0
-        train_dataloader = self.train_store.create_loader(
-            self.config.train.batch_size, eos_token_id=eos_token_id
-        )
-        eval_dataloader = self.eval_pipeline.create_loader(
-            self.config.train.batch_size, shuffle=False
-        )
+        train_dataloader = self.train_store.create_loader(self.config.train.batch_size)
```
Comment (Dahoas) [inline trlx/model/accelerate_ilql_model.py:52]: Where is train_store defined?

File: trlx/model/accelerate_ilql_model.py [hunk 2]
```diff
@@ -107,75 +67,73 @@ def learn(self):
...
             for batch in train_dataloader:
                 if opt_steps % self.config.train.eval_interval == 0:
+                    generate_time = time()
                     self.model.eval()
 
-                    all_samples = []
-                    for prompts in eval_dataloader:
-                        with torch.no_grad():
-                            samples, _ = self.model.sample(
+                    for beta in self.config.method.betas:
```
Comment (Dahoas) [inline trlx/model/accelerate_ilql_model.py:73]: Why do we have multiple betas?

File: trlx/model/accelerate_ppo_model.py [hunk 3]
```diff
@@ -24,8 +24,21 @@
...
         self.store = PPORolloutStorage()
-        super().__init__(config, self.store)
+
+        rollout_loader = self.store.create_loader(
+            self.config.train.batch_size, shuffle=True
+        )
+        self.model, self.opt, self.scheduler, rollout_loader = self.accelerator.prepare(
+            self.model, self.opt, self.scheduler, rollout_loader
+        )
+        self.store.clear_history()
```
Comment (Dahoas) [inline trlx/model/accelerate_ppo_model.py:37]: It turns out when passing a dataloader into deepspeed it is required to be nonempty. I had a hack that loaded a dummy prompt and then clears it once things are loaded. Is this still in the code? I cannot seem to find it

**Comment by LouisCastricato** [comment]
Make sure that you update your code to follow @shahbuland's method of documentation. We also need to update the read the docs after this merge.

File: trlx/model/nn/ilql_models.py [hunk 4]
```diff
@@ -147,35 +149,35 @@ def loss(self, batch):
...
 
         loss_awac = (
             F.cross_entropy(
                 logits[:, :-1].reshape(-1, dsize), actions.reshape(-1), reduction="none"
             ).reshape(bsize, ntokens - 1)
-            * isterminal
+            * terminal_mask
         ).sum() / n_nonterminal
 
         loss = loss_q + loss_v + self.cql_scale * loss_cql + self.awac_scale * loss_awac
```
Comment (Dahoas) [inline trlx/model/nn/ilql_models.py:178]: Do you think it makes more sense to put the loss in the nn.module classes or the accelerator trainer classes? I thought we would want loss defined in the accelerator classes but I'm open to something different if you have a strong opinion?

File: trlx/pipeline/offline_pipeline.py [hunk 5]
```diff
@@ -1,29 +1,39 @@
...
-    def create_loader(self, batch_size: int, shuffle: bool) -> DataLoader:
-        return DataLoader(self, batch_size=batch_size, shuffle=shuffle)
+    def create_loader(self, batch_size: int) -> DataLoader:
+        collate_fn = (
+            DataCollatorWithPadding(self.tokenizer) if self.tokenizer else torch.vstack
+        )
+        return DataLoader(self, batch_size=batch_size, collate_fn=collate_fn)
 
 
 class OfflineRolloutStorage(BaseRolloutStore):
```
Comment (Dahoas) [inline trlx/pipeline/offline_pipeline.py:39]: Perhaps it makes more sense to construct and load our reward labeled datasets in the rollout storage init? I am unsure but I do think the datasets should be separated from the run scripts

File: trlx/trlx.py [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+        )
+        val_prompts = list(map(trim_tokenize, val_prompts))
+        eval_prompts.extend(val_prompts)
+
+    orch = OfflineOrchestrator(
+        model, train_samples=samples, ratings=ratings, metric_fn=metric_fn
+    )
+    model.eval_pipeline = OfflinePipeline(model.tokenizer, eval_prompts)
+
+    model.learn()
```
Comment (Dahoas) [inline trlx/trlx.py:53]: I think we should return the model

Comment (Dahoas) [reply trlx/trlx.py:53]: I like these changes! Having one simple trlx trainer is a good direction.

The readme will need to be updated when this pr is finished.

Before we commit to master make sure we've tested both the ppo and ilql pipelines on the sentiment task.

**Comment by LouisCastricato** [comment]
@dmarx do you mind checking out the architecture choices of this PR?

File: trlx/trlx.py [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+
+from numpy.random import RandomState
+
+from trlx.data.configs import TRLConfig
+from trlx.model.accelerate_ilql_model import AccelerateILQLModel
+from trlx.orchestrator.offline_orchestrator import OfflineOrchestrator
+from trlx.pipeline.offline_pipeline import OfflinePipeline
+
+
+def train(
```
[the remark this answers is not in the document]
Comment (reciprocated) [reply trlx/trlx.py:9]: soon I'll add online as well

File: examples/ilql_randomwalks.py [hunk 8]
```diff
@@ -46,69 +44,54 @@ def generate_random_walks(
...
-    orch = OfflineOrchestrator(
-        model=model,
-        train_samples=train_samples,
-        eval_prompts=eval_prompts,
-        reward_fn=reward_fn,
-        stats_fn=stats_fn,
-    )
-
-    model.learn()
+    trlx.train(walks, lengths, eval_prompts=eval_prompts, metric_fn=metric_fn, config=config, logit_mask=logit_mask)
```
[the remark this answers is not in the document]
Comment (reciprocated) [reply examples/ilql_randomwalks.py:97]: yes, that's the plan

File: examples/ilql_randomwalks.py [hunk 9]
```diff
@@ -46,69 +44,54 @@ def generate_random_walks(
...
+        return {"lengths": lengths,
+                "optimality": (worstlen - bound_lengths) / (worstlen - best_lengths if lengths.shape == best_lengths.shape else 0)}
 
     logit_mask = torch.tensor(~adj)
-    return sample_walks, logit_mask, stats_fn
+    return sample_walks, logit_mask, metric_fn
 
 
 if __name__ == "__main__":
+    walks, logit_mask, metric_fn = generate_random_walks(seed=1000)
```
[the remark this answers is not in the document]
Comment (reciprocated) [reply examples/ilql_randomwalks.py:82]: for use examples I think it's better to go without any subclasses

File: trlx/model/accelerate_ilql_model.py (hunk 1 above)
Comment (reciprocated) [reply trlx/model/accelerate_ilql_model.py:54]: I haven't decided on a proper way for now, so they are defined dynamically

File: trlx/model/accelerate_ilql_model.py (hunk 2 above)
Comment (reciprocated) [reply trlx/model/accelerate_ilql_model.py:75]: to compare against finetune (beta=0)

File: trlx/model/accelerate_ppo_model.py (hunk 3 above)
Comment (reciprocated) [reply trlx/model/accelerate_ppo_model.py:37]: are you talking about clear_history call?

File: trlx/model/accelerate_ppo_model.py [hunk 10]
```diff
@@ -162,6 +174,9 @@ def learn(self, log_fn=None, save_fn=None, eval_fn=None):
                     self.scheduler.step()
                     self.iter_count += 1
 
+                    if self.iter_count % self.config.train.checkpoint_interval == 0:
```
[the remark this answers is not in the document]
Comment (reciprocated) [reply trlx/model/accelerate_ppo_model.py:177]: agree

File: trlx/model/nn/ilql_models.py (hunk 4 above)
Comment (reciprocated) [reply trlx/model/nn/ilql_models.py:192]: I don't have a strong opionion, but forward and generate are also specific here and have to be decoupled and I'm not sure there is a reason for that just yet

File: trlx/model/accelerate_ilql_model.py [hunk 11]
```diff
@@ -107,75 +67,73 @@ def learn(self):
...
+                            logs.update(tensor_stats)
+                            logs.update(mean_metrics)
+
+                            rows = list(zip(samples, *metrics.values()))
+                            logs[f"samples/{beta}"] = wandb.Table(
+                                columns=["samples", *metrics.keys()], rows=rows
+                            )
 
                     self.model.train()
+                    generate_time = time() - generate_time
```
[the remark this answers is not in the document]
Comment (reciprocated) [reply trlx/model/accelerate_ilql_model.py:135]: it doesn't support granular measurements like these

File: examples/ilql_architext.py [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+        "Gives a negative count of rooms for each sample"
+        nrooms = []
+        for s in samples:
+            count = 0.0
+            for char in s:
+                if char == ':':
+                    count -= 1
+            nrooms.append(count)
+
+        return {'nrooms': nrooms}
```
Comment (cat-state) [inline examples/ilql_architext.py:15]: this could be `{ 'nrooms': [-sample.count(':') for sample in samples] }`

**Comment by cat-state** [comment]
Could you make this black formatted & flake8 passing?

**Comment by reciprocated** [comment]
This PR mainly addresses
- Issues #7 and #15 
- It also fixes few inaccuracies in ppo and ilql implementations
- By intertia it also simplifies the readme

**Comment by Dahoas** [comment]
Can you attach wandb runs verifying ppo and ilql still perform appropriately? (I know you have them just wanna make this a standard process)

File: configs/ppo_config.yml [hunk 13]
```diff
@@ -1,52 +1,44 @@
...
-    top_p : 1.0  # top p
-    do_sample : True  # sample
+  name: 'ppoconfig'  # Name of RL method config
+  num_rollouts: 128  # Number of rollouts to collect per epoch
+  chunk_size: 128  # Number of rollouts to collect in one loop of orchestrator
+  ppo_epochs: 4  # Number of ppo epochs
+  init_kl_coef: 0.2  # init kl coefficient
+  target: 6  # target kl coefficient, set None for fixed kl coef
+  horizon: 10000  # PPO horizon
+  gamma: 0.99  # PPO discount
```
Comment (Dahoas) [inline configs/ppo_config.yml:34]: Why the change?

File: trlx/data/ilql_types.py [hunk 14]
```diff
@@ -24,6 +21,9 @@ class ILQLElement:
     input_ids: TensorType["query_size"]
     attention_mask: TensorType["query_size"]
     rewards: TensorType["reward_size"]
+    states_ixs: TensorType["states_size"]
```
Comment (Dahoas) [inline trlx/data/ilql_types.py:24]: What are these for?

File: trlx/model/accelerate_base_model.py [hunk 15]
```diff
@@ -97,44 +96,21 @@ def tokenize(self, text: Iterable[str]):
...
-        self, data: PromptBatch
-    ) -> Tuple[
-        TensorType["chunk_size", "input_length"],
-        TensorType["chunk_size", "gen_size"],
-        Iterable[str],
-    ]:
-        query_tensors = data.tokens.to(
-            self.accelerator.device
-        )  # [B, N] #TODO(dahoas): This may need to be changed
+    def generate(self, input_ids, attention_mask=None, **kwargs):
```
Comment (Dahoas) [inline trlx/model/accelerate_base_model.py:103]: I noticed we are no longer loading the model into accelerate at init time. If we just want to do large model (>20B) inference do we still need to load with accelerate?

File: trlx/model/accelerate_ilql_model.py [hunk 16]
```diff
@@ -1,180 +1,179 @@
-import os
-from typing import Dict, Iterable
+from typing import Iterable, Union
```
Comment (Dahoas) [inline trlx/model/accelerate_ilql_model.py:1]: Unfortunately I really cannot review ilql so I trust things are fine here.

Perhaps it would be a good idea to write some unittests for the ilql implementation moving forward.

File: trlx/model/accelerate_ppo_model.py [hunk 17]
```diff
@@ -91,21 +101,33 @@ def loss(
         advantages = advantages.detach()
 
         all_tokens = torch.cat((query_tensors, response_tensors), dim=1)
-        logits, _, vpred = self.model(all_tokens)
+        attention_mask = (
+            all_tokens.not_equal(self.tokenizer.pad_token_id)
+            .long()
+            .to(all_tokens.device)
+        )
+        position_ids = attention_mask.cumsum(-1) - 1
```
Comment (Dahoas) [inline trlx/model/accelerate_ppo_model.py:109]: Can you leave a comment explaining why this is necessary?

File: trlx/model/accelerate_base_model.py [hunk 18]
```diff
@@ -144,44 +120,144 @@ def get_components(self) -> Dict[str, any]:
         )
         return components
 
+    def save(self, directory=None):
```
Comment (LouisCastricato) [inline trlx/model/accelerate_base_model.py:123]: all of these need doc strings and comments

File: trlx/model/accelerate_ppo_model.py [hunk 19]
```diff
@@ -119,100 +141,42 @@ def loss(
...
-                    "Step: {}, Mean score: {}, pg_loss: {}, vf_loss: {}, kl_coef: {}".format(
-                        self.iter_count,
-                        mean_score,
-                        stats["pg_loss"],
-                        stats["vf_loss"],
-                        self.kl_ctl.value,
-                    )
-                )
-
-    def learn(self, log_fn=None, save_fn=None, eval_fn=None):
```
Comment (Dahoas) [inline trlx/model/accelerate_ppo_model.py:172]: Does the PPO implementation still have multiple ppo_epochs per batch? I see a new variable defining this (n_updated_per_batch) but since we are relying on the base_model's training loop I am not seeing where it gets used.

Perhaps if we do not want to override base_model's learn method we should write something in the post_backward callback

File: trlx/model/nn/ilql_models.py [hunk 20]
```diff
@@ -121,23 +123,23 @@ def loss(self, batch):
             Q = qs[:, :-1].gather(-1, actions).squeeze(-1)
             targetQ = target_qs[:, :-1].gather(-1, actions).squeeze(-1).detach()
 
-        n_nonterminal = max(1, isterminal.sum())
-        V = vs[:, 1:].squeeze() * isterminal
+        n_nonterminal = max(1, terminal_mask.sum())
+        V = vs[:, 1:].squeeze() * terminal_mask
         Q_ = rewards + self.gamma * V
 
         if self.two_qs:
```
[the remark this answers is not in the document]
Comment (Dahoas) [reply trlx/model/nn/ilql_models.py:116]: I agree more comments would be useful

File: trlx/orchestrator/ppo_orchestrator.py [hunk 21]
```diff
@@ -76,14 +78,15 @@ def make_experience(self, num_rollouts: int = 1024, iter_count: int = 0):
...
+
+            ref_logits = ref_logits.to(self.rl_model.accelerator.device)
             logprobs = logprobs_from_logits(logits[:, :-1, :], all_tokens[:, 1:])
             ref_logprobs = logprobs_from_logits(
                 ref_logits[:, :-1, :], all_tokens[:, 1:]
             )
             start = query_tensors.size()[1] - 1
             end = query_tensors.size()[1] + response_tensors.size()[1] - 1
-            all_values = v[:, start - 1 : end - 1]
+            all_values = v[:, start:end]
```
Comment (Dahoas) [inline trlx/orchestrator/ppo_orchestrator.py:89]: I agree this is correct

File: trlx/model/accelerate_base_model.py [hunk 22]
```diff
@@ -144,44 +120,144 @@ def get_components(self) -> Dict[str, any]:
...
+
+            if isinstance(samples[0], str):
+                columns_data = [samples]
+            else:
+                columns_data = [samples.tolist()]
+            columns = ["samples"]
+
+            if self.reward_fn:
+                rewards = torch.as_tensor(self.reward_fn(samples), dtype=torch.float)
+                mean_reward = rewards.mean()
```
Comment (LouisCastricato) [inline trlx/model/accelerate_base_model.py:168]: This needs comments

File: trlx/model/accelerate_base_model.py [hunk 23]
```diff
@@ -144,44 +120,144 @@ def get_components(self) -> Dict[str, any]:
...
+                        self.save()
+                        return self.evaluate()
+
+                self.post_backward_callback()
+
+            self.post_epoch_callback()
+
     @abstractmethod
-    def get_arch(config: TRLConfig):
+    def get_arch(self, config: TRLConfig):
```
Comment (LouisCastricato) [inline trlx/model/accelerate_base_model.py:250]: doc strings

File: trlx/model/accelerate_ilql_model.py [hunk 24]
```diff
@@ -1,180 +1,179 @@
...
-            if self.train_mode
-            else {"model": self.model}
+        input_ids = list(map(torch.as_tensor, tokenized.input_ids))
+        return input_ids
+
+    def post_backward_callback(self):
+        if self.iter_count % self.config.method.steps_for_target_q_sync == 0:
+            self.accelerator.unwrap_model(self.model).sync_target_q_heads()
+
+    def loss(self, batch):
```
Comment (LouisCastricato) [inline trlx/model/accelerate_ilql_model.py:49]: comments

File: trlx/model/accelerate_base_model.py [hunk 25]
```diff
@@ -144,44 +120,144 @@ def get_components(self) -> Dict[str, any]:
...
-    def post_backward_callback(iter, batch, rewards):
+    def post_backward_callback(self):
         pass
 
     @abstractmethod
-    def post_epoch_callback(iter, batch, rewards):
-        """
-        Additional exploration can happen here
-        """
+    def post_epoch_callback(self):
```
Comment (LouisCastricato) [inline trlx/model/accelerate_base_model.py:262]: End in new line

File: trlx/model/accelerate_ilql_model.py [hunk 26]
```diff
@@ -1,180 +1,179 @@
...
+        n_nonterminal = max(1, terminal_mask.sum())
+
+        # values of current states
+        V = vs[:, :-1].squeeze()
+        # values of next states
+        Vnext = vs[:, 1:].squeeze() * dones[:, 1:]
+        # target to fit Q
+        Q_ = rewards + self.model.gamma * Vnext.detach()
+
+        if self.model.two_qs:
```
Comment (LouisCastricato) [inline trlx/model/accelerate_ilql_model.py:88]: Explain what this does.

File: trlx/model/accelerate_ilql_model.py [hunk 27]
```diff
@@ -1,180 +1,179 @@
...
+
+        loss_v = (
+            (
+                (targetQ >= V).int() * self.model.tau * (targetQ - V).pow(2)
+                + (targetQ < V).int() * (1 - self.model.tau) * (targetQ - V).pow(2)
+            )
+            * terminal_mask
+        ).sum() / n_nonterminal
+
+        if self.model.two_qs:
```
Comment (LouisCastricato) [inline trlx/model/accelerate_ilql_model.py:105]: We *cannot* merge this without comments explaining how this is working...

File: trlx/model/accelerate_ppo_model.py [hunk 28]
```diff
@@ -61,15 +62,24 @@ def __init__(self, config, train_mode=True):
...
     def get_arch(self, config: TRLConfig):
-        # TODO(dahoas): Assumes model is gpt like
         return GPTHydraHeadWithValueModel(
             self.config.model.model_path, self.config.model.num_layers_unfrozen
         )
 
-    def loss(
-        self, query_tensors, response_tensors, all_logprobs, all_values, all_rewards
-    ):
+    def loss(self, batch):
```
Comment (LouisCastricato) [inline trlx/model/accelerate_ppo_model.py:76]: Comments

File: trlx/pipeline/ppo_pipeline.py [hunk 29]
```diff
@@ -93,31 +35,34 @@ def create_loader(
...
-                torch.stack(
-                    [elem.query_tensor for elem in elems]
-                ),  # Assumes token tensors all same size
-                torch.stack([elem.response_tensor for elem in elems]),
-                torch.stack([elem.logprobs for elem in elems]),
-                torch.stack([elem.values for elem in elems]),
-                torch.stack([elem.rewards for elem in elems]),
+            return PPORLBatch(
+                # Left padding of already left-padded queries
+                pad_sequence(
```
Comment (Dahoas) [inline trlx/pipeline/ppo_pipeline.py:42]: This is a bit funny but I suppose necessary

File: trlx/model/nn/ilql_models.py [hunk 30]
```diff
@@ -213,70 +159,93 @@ def sync_target_q_heads(self):
         else:
             self._sync_target_q_heads(self.alpha)
 
-    @torch.inference_mode()
-    def sample(
+    def generate(
```
Comment (LouisCastricato) [inline trlx/model/nn/ilql_models.py:162]: Doc string

File: trlx/model/nn/ppo_models.py [hunk 31]
```diff
@@ -389,8 +371,8 @@ def forward_hydra(self, input_ids, **x):
     def forward(
```
Comment (LouisCastricato) [inline trlx/model/nn/ppo_models.py:371]: Doc string

File: trlx/orchestrator/offline_orchestrator.py [hunk 32]
```diff
@@ -1,39 +1,61 @@
...
 @register_orchestrator
 class OfflineOrchestrator(Orchestrator):
-    def __init__(
-        self, model: BaseRLModel, train_samples, eval_prompts, reward_fn, stats_fn=None
-    ):
+    def __init__(self, model, split_token=None):
         self.model = model
+        self.split_token = split_token
+
+    def make_experience(self, samples, rewards):
```
Comment (LouisCastricato) [inline trlx/orchestrator/offline_orchestrator.py:13]: Doc string and comments

File: trlx/orchestrator/ppo_orchestrator.py [hunk 33]
```diff
@@ -20,17 +16,17 @@ class PPOOrchestrator(Orchestrator):
     def __init__(
```
Comment (LouisCastricato) [inline trlx/orchestrator/ppo_orchestrator.py:16]: Doc string

File: trlx/pipeline/ppo_pipeline.py [hunk 34]
```diff
@@ -93,31 +35,34 @@ def create_loader(
...
+                    [elem.query_tensor.flip(0) for elem in elems],
+                    padding_value=self.pad_token_id,
+                    batch_first=True,
+                ).flip(1),
+                # Right pad the rest, to have a single horizontal query/response split
+                pad_sequence(
```
Comment (Dahoas) [inline trlx/pipeline/ppo_pipeline.py:48]: Ok so seeing this now I assume the padded values are handled in loss computation by the attention mask?

**Comment by LouisCastricato** [comment]
The example in the readme is weird... What is it supposed to do? also the simulcra example is kinda odd too.. no explanation of what it is supposed to do.

**Comment by LouisCastricato** [comment]
Looks good to me... Ready to merge
